### DISCUSSION TITLE
Some additions to forward mode `Ad_array`

### DISCUSSION THREAD

**Discussion by vlipovac** [pmgbergen/porepy#831] [Ideas]
Hei,

I have some ideas for the `Ad_array` class, which would simplify some algorithms (at least for me 😄 )

1. Slicing

By overloading
`Ad_array.__getitem__`
and
`Ad_array.__getslice__`

we could allow for a more unified treatment of PorePy's custom AD arays and numpy arrays, i.e.
`A = Ad_array(some_vector, some_matrix)`
`B = A[1:10]`
where the slicing would return a row-wise slicing for both value and Jacobian. It would be equivalent to
`B = Ad_array(some_vector[1:10], some_matrix[1:10])`

This is motivated by the fact that PorePy's AD framework vectorizes the DOFs, so row-wise slicing would be DOF-wise slicing.

2. Logical comparison

I am sure this was already discussed at some point, but again motivated by simpler, unified treatment, it would be nice to have logical comparison.
It seems natural to me to provide a comparison for the values of an Ad_array.
`A < B` would be then `A.val < B.val`.

## Why
This would remove many type checks and if clauses in the code, making code compatible for both numpy arrays and Ad_array instances as arguments.

I think its justified to give the `val` priority over `jac`  in both suggestions, considering that `vals` is the numerical representation of whatever quantitiy the `Ad_array` represents, and `jac` is kind of "secondary" (if I may say so).

**Comment by IvarStefansson** [comment]
Interesting ideas! I agree on prioritising val over jac, although I'm not entirely sure what that implies in your first suggestion?

**Reply by vlipovac** [reply]
> Interesting ideas! I agree on prioritising val over jac, although I'm not entirely sure what that implies in your first suggestion?

As I see it, an `Ad_array` can be seen as a vector-valued function and the slicing as above would apply that certain components of that
vector-valued function are accessed, together with all their derivatives (row in Jacobian)

If we think of an `Ad_array` as a representation of a linearized system and its residual, then above slicing is basically an accessing of a row block and its residual.

**Reply by keileg** [reply]
I think this makes sense.

**Comment by vlipovac** [comment]
This is been resolved with the introduction of EquationSystem
